### PR TITLE
Fix relative symlinks

### LINKED ISSUE #100: Symbolic links are broken!
The new `ln` function does not work correctly in some cases:

```sh
# Setup
mkdir test-shelljs-ln
cd test-shelljs-ln
mkdir -p dir1/dir2/
echo file1_contents > dir1/file1.txt

# Create a link the old-fashioned way
ln -s ../file1.txt dir1/dir2/link1

# Create a link the new way
npm install 'shelljs@git://github.com/arturadib/shelljs.git#284fc1cc34'
node -e "require('shelljs').ln('-s', '../file1.txt', 'dir1/dir2/link2');"
# shelljs prints this error message:
#ln: Source file does not exist

# Inspect the results
ls -la dir1/dir2/
# output contains the following:
#lrwxrwxrwx 1 me me   12 Dec 27 22:47 link1 -> ../file1.txt
#lrwxrwxrwx 1 me me   32 Dec 27 22:49 link2 -> /home/me/file1.txt

# Clean up
cd ..
rm -r test-shelljs-ln
```

So, there are a couple of problems here:

 * The source of a symbolic link doesn't have to exist relative to the current directory.
 * The source of a symbolic link shouldn't be resolved relative to the current directory.  If you really want to check for the existence of the source file, resolve the  source path relative to the **destination** directory, but POSIX `ln` doesn't do this.

The second issue prevents relative links from being created.  The `link2` created by shelljs above points to a file that does not exist.

### PR DESCRIPTION
**Pull request by freitagbr**
Fixes #100

### CODE REVIEW

**Comment by freitagbr** [comment]
I based this PR on Schoonology's fix (PR #235), removing the warning that is displayed when a symbolic link is created.

**Comment by nfischer** [comment]
@freitagbr Could you check that this works under node v0.10? It seems to be failing on the CI.

**Comment by nfischer** [comment]
Also, would you be able to squash the commits? See [here](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/)

**Comment by freitagbr** [comment]
The Node.js v0.10 `path` module does not have the `isAbsolute` method (it was added in Node v0.11). That is what is causing the build to fail.

**Comment by nfischer** [comment]
The usual method we use is to  check for its existence, and if it doesn't exist, use a workaround. Would that be feasible?

**Comment by freitagbr** [comment]
As long as the source string does not start with a `/`, it is absolute. Simply checking `source.indexOf('/') !== 0` should do the same thing, correct?

**Comment by nfischer** [comment]
Will that play well with Windows? I think that should be sufficient for *nix though (unless `~` is supported)

**Comment by freitagbr** [comment]
This should work as intended, though `~` is still not supported. It is trivial to add support, and I think it would be worthwhile. Thoughts?

**Comment by nfischer** [comment]
I'm working on a PR for the tilde (currently unsupported), but there are some other big changes I want to get merged first. So don't add support for it just yet. Once we do, we'll have to add an "is absolute path" method that handles the tilde

File: test/ln.js [hunk 1]
```diff
@@ -84,7 +84,7 @@ assert.equal(
 );
 
 // Abspath regression
-shell.ln('-sf', 'tmp/file1', path.resolve('tmp/abspath'));
+shell.ln('-sf', 'file1', path.resolve('tmp/abspath'));
```
Comment (nfischer) [inline test/ln.js]: :+1:

File: test/ln.js [hunk 2]
```diff
@@ -96,4 +96,21 @@ assert.equal(
...
+assert.equal(
+  fs.readFileSync('tmp/new/file1.txt').toString(),
+  fs.readFileSync('tmp/new/file2.txt').toString()
+);
+fs.writeFileSync('tmp/new/file1.txt', 'new content txt');
+assert.equal(
+  fs.readFileSync('tmp/new/file2.txt').toString(),
+  'new content txt'
+);
+
```
Comment (nfischer) [inline test/ln.js]: Could you please add a test that uses hard links? the semantics for hard links are a little different.

```
$ mkdir -p tmp/
$ echo 'not in tmp' > ./file1.txt
$ echo 'within tmp' > ./tmp/file1.txt
$ ln -sf file1.txt tmp/softlink.txt
$ ln -f file1.txt tmp/hardlink.txt
$ cat tmp/softlink.txt
within tmp
$ cat tmp/hardlink.txt
not in tmp
```

**Comment by nfischer** [comment]
Could you also add some invalid test cases toward the top?

```
# symlinks
mkdir -p tmp/
touch tmp/file1.txt
ln -sf tmp/file1.txt tmp/file2.txt
# make sure there was an error

mkdir -p tmp/
touch file1.txt
ln -sf file1.txt tmp/file2.txt
# make sure there was an error

# hardlinks
mkdir -p tmp/
touch tmp/file1.txt
ln -f file1.txt tmp/file2.txt
# make sure there was an error
```

**Comment by ariporad** [comment]
@freitagbr: Any chance I could ask you to add the small changes? I'd really like to include this in the next release. Thanks!

**Comment by freitagbr** [comment]
@ariporad Will do! I'll have this done by the end of the day.

**Comment by nfischer** [comment]
:+1: @freitagbr just ping this thread when you have the tests merged in and I'll review

**Comment by ariporad** [comment]
@freitagbr: Thanks, but no rush!


Ari

File: test/ln.js (hunk 2 above)
Comment (freitagbr) [inline test/ln.js]: Looks like there already is a test for hard links, line 62 in `test/ln.js`.

Comment (nfischer) [inline test/ln.js]: :+1:

**Comment by nfischer** [comment]
@freitagbr then my only other comment is the invalid test cases (https://github.com/shelljs/shelljs/pull/282#issuecomment-174239702). Once those are in, this'll be good!

File: src/ln.js [hunk 3]
```diff
@@ -44,8 +42,14 @@ function _ln(options, source, dest) {
...
 
   if (options.symlink) {
+    if ((isAbsolute && !fs.existsSync(sourcePath)) || !fs.existsSync(path.resolve(process.cwd(), path.dirname(dest), source))) {
+      common.error('Source file does not exist', true);
+    }
     fs.symlinkSync(source, dest, os.platform() === "win32" ? "junction" : null);
   } else {
+    if (!fs.existsSync(source)) {
+      common.error('Source file does not exist', true);
+    }
```
Comment (freitagbr) [inline src/ln.js]: In testing, I found that the source file was not checked properly for hard links.This correctly checks if the source exists for hard links.

**Comment by nfischer** [comment]
@freitagbr nice work!

LGTM

**Comment by ariporad** [comment]
LGTM!

**Comment by ariporad** [comment]
:+1:

**Pull request merged**